src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -43,6 +45,153 @@ void Bootstrap(const MatType& dataset,
     bootstrapWeights = weights.cols(indices);
 }
 
+/**
+ * Default bootstrap strategy that uses the ::Bootstrap() function. 
+ */
+template<bool UseWeights>
+class DefaultBootstrap
+{
+ public:
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    mlpack::Bootstrap<UseWeights>(dataset,
+                                  labels,
+                                  weights,
+                                  bootstrapDataset,
+                                  bootstrapLabels,
+                                  bootstrapWeights);
+  }
+};
+
+/**
+ * No-op bootstrap strategy that just copies the input to the output.
+ */
+template<bool UseWeights>
+class IdentityBootstrap
+{
+ public:
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    bootstrapDataset = dataset;
+    bootstrapLabels = labels;
+    if (UseWeights)
+      bootstrapWeights = weights;
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ */
+template<bool UseWeights, typename G = std::mt19937>
+class SequentialBootstrap
+{
+ public:
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
```

Comment:
Is there any need to mark this `explicit`?  I think it is superfluous.

Reply:
Also, the documentation implies that `SequentialBootstrap` can just be used as a template parameter and no instantiated bootstrap object is necessary.  But there is no default constructor.  Is the `indicatorMatrix` parameter really necessary?  Or can we automatically compute one?  I don't follow the meaning of it because there is not documentation for the parameter or comments throughout (and like I said I don't have access to the paper or the book).

Reply:
The indicatorMatrix cannot be computed on the fly. The information about the labels lifespan is not known at that point in time.

Reply:
The `explicit` is needed because the constructor only takes one non-defaulted argument

Reply:
Can you explain what the `indicatorMatrix` is?  I personally don't understand it (and don't have access to the paper that describes it).  It's pretty important if we provide this to users that it is very easy to understand what `indicatorMatrix` is, and I'm not excited about forcing a user to compute some quantity about their data, especially when they may have a lot of data.  I strongly suspect it is possible to either automatically compute `indicatorMatrix` on-the-fly, or provide a 'default' in the worst case.

(Also, I don't agree that the `explicit` is necessary.  The code will work just fine without it.)

Reply:
`explicit` reasoning: [C.46: By default, declare single-argument constructors explicit](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit)

We can provide a default for the indicator matrix that will make the SequentialBootstrap behave like the DefaultBootstrap. To compute the indicator matrix many more parameters are needed. Let me explain later when I have more time to elaborate.

Reply:
Yes, I'm glad you pointed that out; I wonder if conversions were the reason you wanted it `explicit`.  In our case, we actually *should* want an implicit conversion here (and in similar situations throughout the mlpack codebase).  A user may pass an Armadillo expression, which is not a `MatType` but is explicitly converted to a `MatType`.  For instance, `arma::zeros<arma::mat>(10, 10)` does not return an `arma::mat`; it returns an `arma::Gen` object indicating that a matrix of zeros should be created.  That in turn may be used inside of larger expressions, represented as types, which can then be optimized at compile-time.

Whenever you have a chance to explain the indicator matrix I would appreciate it, but no hurry.

Reply:
To explain the indicator matrix I drew a simple example of a price series over time.

![image](https://github.com/user-attachments/assets/2bcc683c-9f67-400d-bc1d-9de9f06889f6)

The resulting indicator matrix for this simple example looks like:

```
M = [0 1 1 1 1 1 1 1 1 1 1 1 1 0;
     0 0 0 0 0 1 1 0 0 0 0 0 0 0]
```

We have two events one running from t1 to t12, the other from t5 to t6. During t5, t6 both events are concurrent.

If we just take the data, then we cannot easily derive the indicator matrix as it depends on profit-taking, stop-loss, time-out limits that the user wants to set. In this scenario we assume that at t5 no stop loss is triggered for the first event. Instead, we buy in and start another event concurrent with the first one. You can also question whether you already want to buy in at t1. Why not wait until t6. Or why not have a single event only from t5 to t12.

The concurrency is calculated like this

```
C = [0 1 1 1 1 2 2 1 1 1 1 1 1 0]
```

The uniqueness of each event is then calculated like this:

```
U = [0 1 1 1 1 0.5 0.5 1 1 1 1 1 1 0;
     0 0 0 0 0 0.5 0.5 0 0 0 0 0 0 0]
```

The average uniqueness is then

```
A = [11 / 12; 0.5]
```

Reply:
Thanks for the explanation.  I did some other reading too and I believe that I understand the concept.  Let me check my understanding: it seems that in datasets that the sequential bootstrap is useful for, there are multiple events that could be overlapping.  Each event has a start sample and an end sample.

The concurrency is then just the number of events that are "active" for that sample.  (That is, if we are at sample `i`, the concurrency is the count of events such that `event_start <= i <= event_end`).  And the uniqueness of each event is then the sum of the inverse of the concurrencies for each sample in that event.  Finally the average uniqueness is the average uniqueness across events.  Can you tell me if that is correct?

Then it appears that you sample points using the average uniqueness as weights for the samples?

Reply:
Not exactly. The sampling is essentially a sampling with replacement but the probabilities to draw a sample in the next iteration again is based on its average uniqueness and what samples were drawn in previous runs.